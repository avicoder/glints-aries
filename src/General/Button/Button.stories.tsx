import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Flex } from '../../Layout/Flex';
import { Button, Props as ButtonProps } from './Button';
import {
  ArrowBackDoubleIcon,
  ArrowBackIcon,
  ArrowNextDoubleIcon,
  ArrowNextIcon,
  ViewIcon,
} from '../Icon/components';

export default {
  title: 'General/Button',
  component: Button,
  argTypes: {
    startIcon: {
      control: {
        type: 'select',
        options: {
          'Example: ArrowBackIcon': <ArrowBackIcon />,
          'Example: ArrowBackDoubleIcon': <ArrowBackDoubleIcon />,
        },
      },
    },
    endIcon: {
      control: {
        type: 'select',
        options: {
          'Example: ArrowBackIcon': <ArrowNextIcon />,
          'Example: ArrowNextDoubleIcon': <ArrowNextDoubleIcon />,
        },
      },
    },
    tag: {
      control: { type: 'text' },
    },
    className: {
      table: {
        disable: true,
      },
    },
    ref: {
      table: {
        disable: true,
      },
    },
    theme: {
      control: null,
    },
  },
} as Meta;

const Template: Story<ButtonProps> = args => <Button {...args}>Button</Button>;

export const SolidBlue = Template.bind({});
SolidBlue.args = {
  variant: 'solid-blue',
};

export const Variants: Story<ButtonProps> = () => (
  <Flex flexWrap="wrap" style={{ gap: '10px' }}>
    <Button variant="solid-blue">solid-blue</Button>
    <Button variant="ghost">ghost</Button>
    <Button variant="white-grey">white-grey</Button>
    <Button variant="yellow">yellow</Button>
    <Button variant="solid-white">solid-white</Button>
    <Button variant="link">link</Button>
  </Flex>
);

export const Sizes: Story<ButtonProps> = () => (
  <Flex alignItems="flex-end" flexWrap="wrap" style={{ gap: '10px' }}>
    <Button variant="solid-blue" small>
      Small
    </Button>
    <Button variant="solid-blue">Default</Button>
    <div style={{ width: '100%' }}>
      <Button variant="solid-blue" block small>
        Small Block
      </Button>
    </div>
    <div style={{ width: '100%' }}>
      <Button variant="solid-blue" block>
        Default Block
      </Button>
    </div>
  </Flex>
);

export const Disable: Story<ButtonProps> = () => (
  <Flex style={{ gap: '10px' }}>
    <Button variant="solid-blue">enabled</Button>
    <Button variant="solid-blue" disabled>
      disabled
    </Button>
  </Flex>
);

export const WithIcon: Story<ButtonProps> = () => (
  <Flex style={{ gap: '10px' }}>
    <Button variant="solid-blue" startIcon={<ViewIcon />}>
      button icon left
    </Button>
    <Button variant="solid-blue" endIcon={<ArrowNextIcon />}>
      button icon right
    </Button>
  </Flex>
);

export const WithDifferentTag: Story<ButtonProps> = () => (
  <Flex style={{ gap: '10px' }}>
    <Button variant="solid-blue" tag="a">
      button as anchor
    </Button>
  </Flex>
);
