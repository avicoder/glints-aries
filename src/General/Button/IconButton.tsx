import * as React from 'react';
import classNames from 'classnames';

import { IconBtn } from '../../Style/General/ButtonStyle';

const IconButton: React.FunctionComponent<Props> = ({
  children,
  className,
  shape,
  ...defaultProps
}) => (
  <IconBtn className={classNames('aries-iconbtn', `iconbtn-${shape}`, className)} {...defaultProps}>
    {children}
  </IconBtn>
);

interface Props extends React.ComponentPropsWithoutRef<typeof IconBtn> {
  children: React.ReactNode;
  className: string;
  shape?: string;
  block?: boolean;
}

export default IconButton;
