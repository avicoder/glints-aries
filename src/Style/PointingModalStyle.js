import styled from 'styled-components';
import { PrimaryColor, SecondaryColor } from './Colors';

const padding = '1em 1.5em';

export const PointingModalContainer = styled.div`
    position: relative;
    background: ${SecondaryColor.white};
    box-shadow: 0 10px 15px 0 rgba(0,0,0,0.05);

    &:before {
        position: absolute;
        right: 10px;
        top: -15px;
        content: "";
        border-top: 0 solid transparent;
        border-right: 15px solid transparent;
        border-bottom: 15px solid ${SecondaryColor.white};
        border-left: 15px solid transparent;
    }
`;

export const PointingModalHeaderContainer = styled.header`
    display: flex;
    align-items: center;
    padding: ${padding};

    > *:first-child {
        margin-left: 0;
    }

    > *:last-child {
        margin-right: 0;
    }

    > * {
        margin-left: .5em;
        margin-right: .5em;
    }

    label {
        font-weight: 700;
    }

    span {
        font-size: 12px;
        font-weight: 500;
    }

    p {
        font-size: 16px;
    }
`;

export const PointingModalBodyContainer = styled.div`
    position: relative;
    display: flex;
    flex-direction: column;
    padding: ${padding};
`;

export const PointingModalBodyItemWrapper = styled.p`
    font-weight: 600;
    cursor: pointer;
    margin: .3em 0;

    &:first-child {
        margin-top: 0;
    }

    &:last-child {
        margin-bottom: 0;
    }

    * {
        cursor: pointer;
    }
`;

export const PointingModalFooterContainer = styled.footer`
    display: flex;
    position: relative;
    padding: ${padding};
`;
