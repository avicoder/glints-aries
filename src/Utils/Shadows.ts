export const Shadows = {
  1: {
    up: '0 2px 4px 0 rgba(0, 0, 0, 0.15)',
  },
  2: {
    up: '0 4px 8px 0 rgba(0, 0, 0, 0.15)',
  },
  3: {
    up: '0 6px 12px 0 rgba(0, 0, 0, 0.10)',
  },
  4: {
    up: '0 12px 24px 0 rgba(0, 0, 0, 0.10)',
    down: '0 -6px 12px 0 rgba(0, 0, 0, 0.10)',
  },
};
