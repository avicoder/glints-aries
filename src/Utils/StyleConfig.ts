export enum ETabAlignment {
  HORIZONTAL = 'horizontal',
  VERTICAL = 'vertical',
}

export enum EHorizontalTabVariant {
  UNDERLINED = 'underlined',
  COLORED = 'colored',
}

export enum ETooltipPosition {
  TOP = 'top',
  BOTTOM = 'bottom',
  LEFT = 'left',
  RIGHT = 'right',
}

export enum Variant {
  SOLID = 'solid',
  GHOST = 'ghost',
  LINK = 'link',
  SOLID_SHADOW = 'solid-shadow',
}

export const PsychedelicTheme = {
  BLUE_DOMINANT: 'blue-dominant',
  BLUE_DEFAULT: 'blue-default',
  RED_DOMINANT: 'red-dominant',
  RED_DEFAULT: 'red-default',
  YELLOW_DOMINANT: 'yellow-dominant',
  YELLOW_DEFAULT: 'yellow-default',
};

export enum Theme {
  RED = 'red',
  YELLOW = 'yellow',
  BLUE = 'blue',
  BLUE_RED = 'blue-red',
  WHITE = 'white',
  GREY = 'grey',
}

export enum Size {
  SMALL = 'small',
  LARGE = 'large',
}

/* Screen size library of Glints
 * Usage:
 * @media ${Device.mobileL} { ... }
 */

export const ScreenSize = {
  mobileS: 320,
  mobileM: 480,
  mobileL: 640,
  tablet: 768,
  desktopS: 1024,
  desktopM: 1260,
  desktopL: 1440,
};

export const Device = {
  mobileS: `(max-width: ${ScreenSize.mobileS}px)`,
  mobileM: `(max-width: ${ScreenSize.mobileM}px)`,
  mobileL: `(max-width: ${ScreenSize.mobileL}px)`,
  tablet: `(max-width: ${ScreenSize.tablet}px)`,
  desktopS: `(max-width: ${ScreenSize.desktopS}px)`,
  desktopM: `(max-width: ${ScreenSize.desktopM}px)`,
  desktopL: `(max-width: ${ScreenSize.desktopL}px)`,
};
