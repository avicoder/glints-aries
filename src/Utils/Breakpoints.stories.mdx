import { Meta } from '@storybook/addon-docs/blocks';
import { ScreenSize, Device } from './StyleConfig';

<Meta title="Utilities/Breakpoints" />

# Breakpoints

These utilities are used internally by components like
[GlintsContainer](/?path=/docs/utilities-glintscontainer--page#glintscontainer).
You should use them when building your own responsive styles to ensure that
breaks only happen at a few well-defined viewport-widths.

## Device

Our device classes. Names are generally oriented upwards: For example, when you
use `Device.desktopS`, your styles will apply for all viewport-widths greater or
equal to `1024px`. That means that you have to override those styles with rules
for wider viewports if needed. Thus, if you have styles for serveral widths, the
rules for narrow viewports should be at the top and should be consecutively
overridden by rules for wider viewports towards the bottom of your css.

`mobileS` is an exception: It applies "downwards", i.e. on all screens smaller
than `480px`.

```TypeScript
import { Device } from 'glints-aries';

const MyContainer = styled.div`
  // Applies on screens smaller than tablet
  color: red;

  // Applies from tablets to screens smaller than large desktops
  @media ${Device.tablet} {
    color: green;
  }

  // Applies on large desktops
  @media ${Device.desktopL} {
    color: blue;
  }
`;
```

<table>
  <thead>
    <tr>
      <th>Device</th>
      <th>Value</th>
    </tr>
  </thead>
  <tbody>
    {Object.keys(Device).map(dev => (
      <>
        <tr>
          <td>{dev}</td>
          <td>{Device[dev]}</td>
        </tr>
      </>
    ))}
  </tbody>
</table>

## ScreenSize

These are the numeric pixel values that define our breakpoints. The values are
the smallest pixel width that should be classified as the respective constant's
size class. For example, a viewport needs to be at least `1440px` wide to "qualify" as
Desktop Large. Note that the values do not include the `px` prefix;

```TypeScript
import { ScreenSize } from 'glints-aries';

const MyContainer = styled.div`
  color: red;
  @media (min-width: ${ScreenSize.desktopL}px) {
    color: blue;
  }
`;
```

<table>
  <thead>
    <tr>
      <th>Device</th>
      <th>Value</th>
    </tr>
  </thead>
  <tbody>
    {Object.keys(ScreenSize).map(size => (
      <>
        <tr>
          <td>{size}</td>
          <td>{ScreenSize[size]}</td>
        </tr>
      </>
    ))}
  </tbody>
</table>
