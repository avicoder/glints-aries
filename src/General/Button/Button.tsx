import React from 'react';

import SolidButton from './SolidButton';
import SolidShadowButton from './SolidShadowButton';
import GhostButton from './GhostButton';
import LinkButton from './LinkButton';

import { ButtonVariant, ButtonTheme } from '../../Utils/StyleConfig';
import {
  StartIconContainer,
  EndIconContainer,
  WhiteGreyBtn,
} from './ButtonStyle';

const SOLID_BUTTON_THEME_MAP = {
  [ButtonVariant.SOLID_WHITE]: ButtonTheme.WHITE,
  [ButtonVariant.SOLID_BLUE]: ButtonTheme.BLUE,
};

export const transformVariant = (variant: string, theme?: string) => {
  if (theme) {
    console.warn(
      `Warning: Button's theme prop is deprecated and will be removed in a future release.\nPlease use the variant prop instead.`
    );
  }

  if (
    (variant === ButtonVariant.SOLID_WHITE || variant === 'default') &&
    theme &&
    theme !== ButtonTheme.WHITE
  ) {
    console.warn(
      `Warning: Button's 'default' variant is deprecated and will be removed in a future release.\nPlease use the 'solid-blue' or 'solid-white' variant instead.`
    );
    return ButtonVariant.SOLID_BLUE;
  } else if (variant === 'primary') {
    console.warn(
      `Warning: Button's 'primary' variant is deprecated and will be removed in a future release.\nPlease use the 'yellow' variant instead.`
    );
    return ButtonVariant.YELLOW;
  } else if (Object.values(ButtonVariant).includes(variant)) {
    return variant;
  }
  return ButtonVariant.SOLID_WHITE;
};

export const Button: React.FC<Props> = ({
  children,
  block = false,
  className,
  disabled = false,
  onClick,
  small = false,
  variant = ButtonVariant.SOLID_WHITE,
  startIcon,
  endIcon,
  theme,
  forwardedRef,
  ...defaultProps
}) => {
  const content = (
    <>
      {startIcon && <StartIconContainer>{startIcon}</StartIconContainer>}
      {children}
      {endIcon && <EndIconContainer>{endIcon}</EndIconContainer>}
    </>
  );

  const transformedVariant = transformVariant(variant, theme);

  switch (transformedVariant) {
    case ButtonVariant.YELLOW:
      return (
        <SolidShadowButton
          ref={forwardedRef}
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          small={small}
          {...defaultProps}
        >
          {content}
        </SolidShadowButton>
      );
    case ButtonVariant.GHOST:
      return (
        <GhostButton
          ref={forwardedRef}
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          small={small}
          {...defaultProps}
        >
          {content}
        </GhostButton>
      );
    case ButtonVariant.WHITE_GREY:
      return (
        <WhiteGreyBtn
          ref={forwardedRef}
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          small={small}
          {...defaultProps}
        >
          {content}
        </WhiteGreyBtn>
      );
    case ButtonVariant.LINK:
      return (
        <LinkButton
          ref={forwardedRef}
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          {...defaultProps}
        >
          {content}
        </LinkButton>
      );
    default:
      return (
        <SolidButton
          ref={forwardedRef}
          theme={SOLID_BUTTON_THEME_MAP[transformedVariant]}
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          small={small}
          {...defaultProps}
        >
          {content}
        </SolidButton>
      );
  }
};

export type ButtonVariantType =
  | 'solid-white'
  | 'solid-blue'
  | 'yellow'
  | 'ghost'
  | 'white-grey'
  | 'link';

export interface Props extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /** Make the button fit to its parent width. */
  block?: boolean;
  className?: string;
  /** Sets Button to disable state. */
  disabled?: boolean;
  /** Callback when click. */
  onClick?(e: React.MouseEvent<HTMLButtonElement, MouseEvent>): void;
  /** Sets Button to small version. */
  small?: boolean;
  /** The variant of the button. */
  variant?: ButtonVariantType;
  /** Changes the tag with which the button will render. */
  tag?: React.ElementType;
  /** Element placed before the children. */
  startIcon?: React.ReactNode;
  /** Element placed after the children. */
  endIcon?: React.ReactNode;
  /** Deprecated! */
  theme?: string;
  forwardedRef?: React.RefObject<HTMLButtonElement>;
  /** Determines if the button should have rounded corners */
  isRoundedCorner?: boolean;
}

const forwardRef = (
  props: Props,
  forwardedRef: React.RefObject<HTMLButtonElement>
) => <Button forwardedRef={forwardedRef} {...props} />;

forwardRef.displayName = Button.name;

export default React.forwardRef(forwardRef);
