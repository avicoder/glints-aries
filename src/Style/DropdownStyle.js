import styled, { keyframes } from 'styled-components';
import { PrimaryColor, SecondaryColor } from './Colors';
import { Size } from '../Utils/StyleConfig';

const growing = keyframes`
    from {
        max-height: 0;
    }
    to {
        max-height: auto;
    }
`;

export const DropdownLabelWrapper = styled.div`
    display: flex;
    justify-content: space-between;
    align-items: center;
    cursor: pointer;
`;

export const DropdownLabel = styled.div`
    margin-right: 3em;
    cursor: pointer;
`;

export const DropdownItemWrapper = styled.div`
    position: absolute;
    width: 100%;
    background: ${SecondaryColor.white};
    padding: .4em 1em;
    box-shadow: 0 10px 15px 0 rgba(0,0,0,0.05);
    max-height: 180px;
    overflow-y: auto;
    animation: ${growing} .2s linear;
`;

export const DropdownItem = styled.div`
    display: flex;
    cursor: pointer;
    margin-bottom: .5em;
    color: ${SecondaryColor.lightblack};
    

    &:last-child {
        margin-bottom: 0;
    }

    &:hover {
        color: ${PrimaryColor.glintsblue};
    }
`;

export const DropdownContainer = styled.div`
    position: relative;
    outline: none;

    ${DropdownLabelWrapper} {
        background: ${props => props.open ? `${SecondaryColor.blue}` : `${SecondaryColor.white}`};
        color: ${props => props.open ? `${SecondaryColor.white}` : `${SecondaryColor.blue}`};
        transition: all .3s;

        &:hover {
            background: ${SecondaryColor.blue};
            color: ${SecondaryColor.white};
            transition: all .3s;
        }
    }
    
    ${DropdownLabel} {
        font-size: ${props => props.size === `${Size.SMALL}`
    ? '12px'
    : props.size === `${Size.MEDIUM}`
      ? '16px'
      : props.size === `${Size.LARGE}`
        ? '20px'
        : '14px'};
    }

    ${DropdownItem} {
        font-size: ${props => props.size === `${Size.SMALL}`
    ? '10px'
    : props.size === `${Size.MEDIUM}`
      ? '12px'
      : props.size === `${Size.LARGE}`
        ? '14px'
        : '12px'};
    }
`;
