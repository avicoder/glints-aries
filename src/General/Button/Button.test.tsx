import * as React from 'react';
import Button from '.';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';

import { Variant } from '../../Utils/StyleConfig';
import { PrimaryColor, SecondaryColor, Greyscale } from '../../Style/Colors';

it('<Button> should render with text "click me" and an onClick handler', () => {
  const onClick = jest.fn();
  const ButtonSnapshot = renderer
    .create(<Button onClick={onClick}>click me</Button>)
    .toJSON();
  expect(ButtonSnapshot).toMatchSnapshot();
});

it('<Button> should render as button by default', () => {
  const { getByTestId } = render(
    <Button data-testid="test-button">test</Button>
  );
  expect(getByTestId('test-button').tagName).toBe('BUTTON');
});

describe('<Button> should render as the given tag', () => {
  const tags = ['a', 'h1', 'div'];
  tags.forEach((tag: string) => {
    it(`<Button> should render as ${tag}`, () => {
      const { container } = render(
        <Button tag={tag as React.ElementType}>button</Button>
      );
      const tagRegexp = new RegExp(tag, 'i');
      expect(container.querySelector(tag).tagName).toMatch(tagRegexp);
    });
  });
});

it('should call onClick once when it is clicked', () => {
  const onClick = jest.fn();
  const { getByText } = render(<Button onClick={onClick}>click me</Button>);
  const button = getByText('click me');
  fireEvent.click(button);
  expect(onClick).toHaveBeenCalledTimes(1);
});

it('should not call onClick when it is disabled', () => {
  const onClick = jest.fn();
  const { getByText } = render(
    <Button onClick={onClick} disabled>
      click me
    </Button>
  );
  const button = getByText('click me');
  fireEvent.click(button);
  expect(onClick).toHaveBeenCalledTimes(0);
});

describe('contains the correct className for each variant:', () => {
  Object.values(Variant).forEach(variant => {
    it(`${variant}`, () => {
      const { getByText } = render(<Button variant={variant}>click me</Button>);
      const button = getByText('click me');
      expect(button).toHaveClass(`aries-${variant}btn`);
    });
  });
});

describe('render the correct default style for each variant:', () => {
  const defaultStyleMap = {
    [Variant.SOLID]: {
      bgColor: Greyscale.white,
      color: SecondaryColor.actionblue,
    },
    [Variant.GHOST]: {
      bgColor: Greyscale.white,
      color: SecondaryColor.actionblue,
    },
    [Variant.SOLID_SHADOW]: {
      bgColor: PrimaryColor.glintsyellow,
      color: Greyscale.black,
    },
    [Variant.LINK]: {
      bgColor: Greyscale.white,
      color: SecondaryColor.actionblue,
    },
  };

  Object.values(Variant).forEach(variant => {
    it(`${variant}`, () => {
      const defaultStyle = defaultStyleMap[variant];
      const { getByText } = render(<Button variant={variant}>click me</Button>);
      const button = getByText('click me');
      expect(button).toHaveStyle(`background-color: ${defaultStyle.bgColor}`);
      expect(button).toHaveStyle(`color: ${defaultStyle.color}`);
    });
  });
});

describe('render the correct disabled style for each variant:', () => {
  Object.values(Variant).forEach(variant => {
    it(`${variant}`, () => {
      const { getByText } = render(
        <Button variant={variant} disabled>
          click me
        </Button>
      );
      const button = getByText('click me');
      expect(button).toHaveStyle(`background-color: ${Greyscale.lightgrey}`);
      expect(button).toHaveStyle(`color: ${Greyscale.white}`);
    });
  });
});
