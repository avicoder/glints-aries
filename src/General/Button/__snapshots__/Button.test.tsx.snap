// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> should render with text "click me" and an onClick handler 1`] = `
<button
  className="ButtonStyle__Button-gw4ytw-0 ButtonStyle__SolidButton-gw4ytw-1 dwMShH aries-solidbtn"
  onClick={[MockFunction]}
>
  click me
</button>
`;
